### ISSUE TITLE
Feature visualization

### ISSUE THREAD

**Issue by yangershuai627** [ultralytics/ultralytics#20207]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Does YOLOv8 provide a visual script for feature mapping to analyze the current model training? For example, heat map visualization and Grad-Based CAM visualization.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yangershuai627, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response. An Ultralytics engineer will review your question and assist you soon 😊

**Comment by glenn-jocher** [comment]
Hi @yangershuai627,

Yes, Ultralytics provides feature map visualization through the `feature_visualization` function in the `utils/plotting.py` module. This function allows you to visualize feature maps during inference and save them as PNG images.

For heatmap visualization, Ultralytics offers a dedicated Heatmap solution that can visualize object movement patterns and density. This is available through both CLI and Python API:

```python
# CLI
yolo solutions heatmap show=True

# Python
from ultralytics import solutions
heatmap = solutions.Heatmap(model="yolo11n.pt", colormap=cv2.COLORMAP_PARULA)
```

Regarding Grad-CAM visualization, while not built-in directly, you could implement it using the model's hooks to access activations and gradients. Would you like specific guidance on implementing Grad-CAM with YOLO models?

**Comment by yangershuai627** [comment]
@Y-T-G @glenn-jocher Thank you very much for your reply!
However, this method can only obtain the heat map of this image, as shown in the following figure. 

![Image](https://github.com/user-attachments/assets/476e900c-57cf-4c55-b6a0-cfa14ee510ac)

How can the heat maps of the three scales output by the backbone or neck of the yolov8 model be obtained?

**Comment by glenn-jocher** [comment]
Hi @yangershuai627,

You can visualize feature maps from different scales (backbone or neck) by setting `visualize` to a path instead of just `True`. For example:

```python
from ultralytics import YOLO
import os

# Create directory for visualizations
save_dir = os.path.join("runs", "detect", "feature_maps")
os.makedirs(save_dir, exist_ok=True)

# Load model and predict with visualization
model = YOLO("yolov8n.pt")
results = model.predict("path/to/image.jpg", visualize=save_dir)
```

This will save feature maps from all layers to the specified directory with names like `stage{layer_index}_{layer_type}_features.png`. The backbone layers are typically in the early stages, and neck layers (like FPN) are in the middle stages before the detection head.

You can then identify which specific stage numbers correspond to the three scales you're interested in by examining the saved visualizations.

**Comment by yangershuai627** [comment]
@glenn-jocher This is very helpful, Thank you!

**Comment by glenn-jocher** [comment]
You're welcome @yangershuai627! I'm glad the feature map visualization solution was helpful for your needs. If you encounter any other questions while working with the feature visualizations or other aspects of YOLO models, feel free to ask. The community and our team are always here to help.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #25548](https://github.com/ultralytics/ultralytics/pull/25548), titled **“Replace predict feature-map dumps with class activation heatmaps.”**

Key changes include:

- Replaced raw layer-by-layer feature-map dumps with practical PyTorch class activation heatmaps using LayerCAM.
- Updated `visualize=True` during prediction to generate heatmap overlays alongside prediction outputs.
- Applied confidence and class filters, with support for visualizing a configurable number of top predictions.
- Limited activation-map generation to supported PyTorch workflows, excluding tensor sources, augmentation, and embedding workflows.
- Updated inference and documentation to reflect the new heatmap behavior.

These changes should provide more interpretable visual explanations by highlighting image regions that influence predictions, while ensuring visualizations focus on relevant detections.

The changes were implemented by [@Y-T-G](https://github.com/Y-T-G), with contributions from [@glenn-jocher](https://github.com/glenn-jocher).

To test the latest changes, install directly from the `main` branch:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, you can wait for the next official PyPI release.

Please test this update and share whether it resolves the issue. Thank you 🙏 for reporting it, and any further feedback is welcome if the problem persists.

**Comment by glenn-jocher** [comment]
@yangershuai627, #25548 by @Y-T-G is now merged and replaces the previous raw feature-map visualization with a Grad-based LayerCAM view. `visualize=True` produces one compact heatmap per image showing the pixels that increased the selected prediction scores.
